ldap3/strategy/base.py

```diff
@@ -194,6 +194,15 @@ def _open_socket(self, address, use_ssl=False, unix_socket=False):
         Tries to open and connect a socket to a Server
         raise LDAPExceptionError if unable to open or connect socket
         """
+        def _cleanup_socket():
+            """ Close the open socket. Call before raising exceptions so no
+                lingering sockets are left about.
+            """
+            try:
+                self.connection.socket.close()
+            except OSError:
+                pass
```

Comment:
would this make more sense as an annotation to wrap functions than as a standalone function to be called? then it would check if an exception was being raised before deciding whether to close the socket or not

that way we get the niceties of python context management and it operates more like a try/finally for the entire function

Reply:
Is this what you meant? (i assumed you meant decorator, i'm not sure how to solve this with annotations)
Personally i'm not sure i find this much more readable.
```diff
diff --git a/ldap3/strategy/base.py b/ldap3/strategy/base.py
--- a/ldap3/strategy/base.py
+++ b/ldap3/strategy/base.py
@@ -23,6 +23,7 @@
 # along with ldap3 in the COPYING and COPYING.LESSER files.
 # If not, see <http://www.gnu.org/licenses/>.
 
+import functools
 import socket
 try:  # try to discover if unix sockets are available for LDAP over IPC (ldapi:// scheme)
     # noinspection PyUnresolvedReferences
@@ -189,6 +190,26 @@ class BaseStrategy(object):
         if self.connection.usage:
             self.connection._usage.stop()
 
+    def close_connection_on_exception(orig_func):
+        """
+        A decorator that will close self.connection.socket if the
+        original function errors. Should only be applied to functions that
+        are executed before self.connection.bound and self.connection.closed
+        are set (ie _open_socket)
+        """
+        @functools.wraps(orig_func)
+        def test_wrapper(self, *args, **kwargs):
+            try:
+                return orig_func(*args, **kwargs)
+            except Exception as e:
+                try:
+                    self.connection.socket.close()
+                except Exception:
+                    pass
+                raise
+        return test_wrapper
+
+    @close_connection_on_exception
     def _open_socket(self, address, use_ssl=False, unix_socket=False):
         """
         Tries to open and connect a socket to a Server
```

Reply:
yeah that's what i'd meant. i moreso meant that it was easier for people adding new things to realize that they should use it than easier to read what it does